pipeline_interface.yaml

```diff
@@ -10,35 +10,34 @@ pipelines:
   peppro.py:
     name: PEPPRO
     path: pipelines/peppro.py
+    schema: https://schema.databio.org/pipelines/ProseqPEP.yaml
     looper_args: True
     required_input_files: [read1]
     all_input_files: [read1, read2]
     ngs_input_files: [read1, read2]
-    arguments:
-      "--sample-name": sample_name
-      "--genome": genome
-      "--input": read1
-      "--single-or-paired": read_type
-    optional_arguments:
-      "--input2": read2
-      "--runon": runon
-      "--adapter": adapter
-      "--dedup": dedup
-      "--trimmer": trimmer
-      "--umi": umi
-      "--umi_len": umi_len
-      "--max_len": max_len
-      "--sob": sob
-      "--scale": scale
-      "--parts": parts
-      "--prealignments": prealignments
-      "--TSS-name": TSS_name
-      "--pre-name": pre_name
-      "--anno-name": anno_name
-      "--coverage": coverage
-      "--keep": keep
-      "--noFIFO": no_fifo
-      "--complexity": complexity
+    command_template: >
+      {pipeline.path}
+      --sample-name {sample.sample_name}
+      --genome {sample.genome}
+      --input {sample.read1}
+      --single-or-paired {sample.read_type}
+      {% if sample.read2 is defined %} --input2 {sample.read2} {% endif %}
+      {% if sample.adapter is defined %} --adapter-tool {sample.adapter} {% endif %}
+      {% if sample.dedup is defined %} --dedup-tool {sample.dedup} {% endif %}
+      {% if sample.umi_len is defined %} --umi-len {sample.umi_len} {% endif %}
+      {% if sample.max_len is defined %} --max-len {sample.max_len} {% endif %}
+      {% if sample.sob is defined %} --sob {sample.sob} {% endif %}
+      {% if sample.scale is defined %} --scale {sample.scale} {% endif %}
+      {% if sample.prealignments is defined %} --prealignments {sample.prealignments} {% endif %}
+      {% if sample.TSS_name is defined %} --TSS-name {sample.TSS_name} {% endif %}
+      {% if sample.pre_name is defined %} --pre-name {sample.pre_name} {% endif %}
+      {% if sample.anno_name is defined %} --anno_name {sample.anno_name} {% endif %}
+      {% if sample.search_file is defined %} --search-file {sample.search_file} {% endif %}
+      {% if sample.coverage is defined %} --coverage {sample.coverage} {% endif %}
+      {% if sample.keep is defined %} --keep {sample.keep} {% endif %}
+      {% if sample.no_fifo is defined %} --noFIFO {sample.no_fifo} {% endif %}
+      {% if sample.prioritize is defined %} --prioritize {sample.prioritize} {% endif %}
+      {% if sample.complexity is defined %} --complexity {sample.complexity} {% endif %}
```

Comment:
That's another pure jinja way to encode optional args in command template:

```
      {pipeline.path}
      --sample-name {sample.sample_name}
      --genome {sample.genome}
      --input {sample.read1}
      --single-or-paired {sample.read_type}
      {" --input2 {}".format(sample.read2) if sample.read2 is defined else ""}
      {" --adapter-tool {}".format(sample.adapter) if sample.adapter is defined else ""}
      {" --dedup-tool {}".format(sample.dedup) if sample.dedup is defined else ""}
      {" --umi-len {}".format(sample.umi_len) if sample.umi_len is defined else ""}
      {" --max-len {}".format(sample.max_len) if sample.max_len is defined else ""}
      {" --sob {}".format(sample.sob) if sample.sob is defined else ""}
      {" --scale {}".format(sample.scale) if sample.scale is defined else ""}
      {" --prealignments {}".format(sample.prealignments) if sample.prealignments is defined else ""}
      {" --TSS-name {}".format(sample.TSS_name) if sample.TSS_name is defined else ""}
      {" --pre-name {}".format(sample.pre_name) if sample.pre_name is defined else ""}
      {" --anno-name {}".format(sample.anno_name) if sample.anno_name is defined else ""}
      {" --search-file {}".format(sample.search_file) if sample.search_file is defined else ""}
      {" --coverage {}".format(sample.coverage) if sample.coverage is defined else ""}
      {" --keep {}".format(sample.keep) if sample.keep is defined else ""}
      {" --noFIFO {}".format(sample.no_fifo) if sample.no_fifo is defined else ""}
      {" --prioritize {}".format(sample.prioritize) if sample.prioritize is defined else ""}
      {" --complexity {}".format(sample.complexity) if sample.complexity is defined else ""}
```

it isn't less verbose, but it's more pythonic

Reply:
in that case we could write a function that would permit this:

```
{ " --input2 {}".ifdef(sample.read2) }
```

not saying we *should*...

Reply:
using a possibly undefined attribute as a macro argument would raise an error, since I used [`jinja2.StrictUndefined`](https://jinja.palletsprojects.com/en/2.11.x/api/#jinja2.StrictUndefined) class to catch undefined attributes in the command template (this way jinja takes care of command template validation in sample/project/pipeline context).

As their docs say: 
> you can do nothing with it (undefined attribute) except checking if it’s defined using the defined test

Anyway, these two options work since it's pure jinja, so we do not need to decide now

Reply:
ok perfect.